src/Breadcrumbs.php

```diff
@@ -147,77 +180,166 @@ public function links(array $value): self
+    public function listTagName(string $value): self
     {
         $new = clone $this;
-        $new->tag = $value;
+        $new->listTagName = $value;
 
         return $new;
     }
 
+    public function render(): string
+    {
+        $attributes = $this->attributes;
+        $attributes['aria-label'] ??= 'breadcrumb';
+
+        if ($this->links === []) {
+            return '';
+        }
+
+        $list = $this->renderList();
+
+        if ($list === '') {
+            return '';
+        }
+
+        return Nav::tag()->addAttributes($attributes)->content("\n", $list, "\n")->encode(false)->render();
+    }
+
+    private function renderList(): string
+    {
+        $listAttributes = $this->listAttributes;
+        $classes = $listAttributes['class'] ?? null;
+
+        /** @psalm-var non-empty-string|null $listId */
+        $listId = match ($this->listId) {
+            true => $listAttributes['id'] ?? Html::generateId(self::LIST_NAME . '-'),
+            '', false => null,
+            default => $this->listId,
+        };
+
+        unset($listAttributes['class'], $listAttributes['id']);
+
+        $items = [];
+        $active = 0;
+
+        foreach ($this->links as $link) {
+            $active += (int) $link->isActive();
+            $items[] = $this->renderItem($link);
+
+            if ($active > 1) {
+                throw new RuntimeException('Only one "link" can be active.');
+            }
+        }
+
+        $items = implode("\n", $items);
+
+        if ($items === '') {
+            return '';
+        }
+
+        Html::addCssClass($listAttributes, [self::LIST_NAME, $classes]);
+
+        if ($this->listTagName === '') {
+            throw new InvalidArgumentException('List tag cannot be empty.');
+        }
+
+        return Html::tag($this->listTagName)
+            ->attributes($listAttributes)
+            ->content("\n", $items, "\n")
+            ->id($listId)
+            ->encode(false)
+            ->render();
+    }
+
     /**
      * Renders a single breadcrumb item.
      *
-     * @param array $link the link to be rendered. It must contain the "label" element. The "url" element is optional.
-     * @param string $template the template to be used to rendered the link. The token "{link}" will be replaced by the
-     * link.
-     *
-     * @throws JsonException|RuntimeException if `$link` does not have "label" element.
+     * @param Link $link The breadcrumb item to render.
      *
-     * @return string the rendering result
+     * @return string The rendering result.
      */
-    private function renderItem(array $link, string $template): string
+    private function renderItem(Link $link): string
     {
-        $encodeLabel = ArrayHelper::remove($link, 'encode', $this->encodeLabels);
-
-        if (array_key_exists('label', $link)) {
-            $label = $encodeLabel ? Html::encode($link['label']) : $link['label'];
-        } else {
+        if ($link->label === '') {
```

Comment:
Label can be empty string. For example. appearance set by CSS.

Reply:
Please could you give me an example where this case occurs, what happens if all the labels are empty-string, in the case of the Breadcrumb component in bootstrap5, there is no case with the label empty string.

Reply:
Link with class "myCssClass". CSS:

```css
.myCssClass {
	display: inline-block;
	width: 32px;
	height: 32px;
	background: red url(...);
}
```

Reply:
done.